### PR TITLE
Add visionOS support

### PR DESCRIPTION
**Pull request by stonko1994**
This PR builds upon https://github.com/Quick/Nimble/pull/1083 but includes CocoaPods and testing support.

Before this PR can be merged, two counterparts PRs in [CwlPreconditionTesting](https://github.com/mattgallagher/CwlPreconditionTesting) need to be merged first:
- https://github.com/mattgallagher/CwlPreconditionTesting/pull/33
- https://github.com/mattgallagher/CwlPreconditionTesting/pull/34

_Based on this PR, I opened a discussion to outline a problematic usage of Carthage._

---

Checklist - While not every PR needs it, new features should consider this list:

- [x] Does this have tests?
- [x] Does this have documentation?
- [ ] Does this break the public API (Requires major version bump)?
- [x] Is this a new feature (Requires minor version bump)?

### CODE REVIEW

File: Package@swift-5.9.swift
Comment (stonko1994) [inline Package@swift-5.9.swift:1]: This file was added to not break support for older Swift versions.

See details about this here: https://www.polpiella.dev/version-specific-package-manifests

**Comment by stonko1994** [comment]
@younata I have some questions how to handle certain things in this PR:
- How to handle the dependency PRs of `CwlPreconditionTesting` if they are not getting picked up in a timely manner?
- What is the strategy for the CI and the tests? Currently, it would fail since there is no Xcode 15.1 installed, and therefore, visionOS is unknown. Should we skip the visionOS tests for now and update to the according macOS images once it is included in a stable Xcode version?

**Comment by younata** [comment]
This is good work! Thank you so much for pushing for this.

> - How to handle the dependency PRs of CwlPreconditionTesting if they are not getting picked up in a timely manner?

I think what we have here, pointing to a fork, is good for now. I'm hesitant to release a version pointing to the forks, but we'll cross that bridge when we get there. Presumably CwlPreconditionTesting is also hesitant to officially support for an unreleased platform.

> - What is the strategy for the CI and the tests? Currently, it would fail since there is no Xcode 15.1 installed, and therefore, visionOS is unknown. Should we skip the visionOS tests for now and update to the according macOS images once it is included in a stable Xcode version?

I think that's the best we can do for now. We'll probably have to wait even longer for GitHub to add images that include visionOS.

**Comment by stonko1994** [comment]
Thanks 🙇‍♂️ I modified the test script to skip visionOS tests if no SDK is present: [b86604f](https://github.com/Quick/Nimble/pull/1098/commits/b86604f0bc06d55c495d980e5aa1d55217cdd5db)

**Comment by stonko1994** [comment]
@younata, I tried to fix the failing CI steps. Should be working now.

**Comment by stonko1994** [comment]
The PRs in [CwlPreconditionTesting](https://github.com/mattgallagher/CwlPreconditionTesting) were merged and I reverted back to the official repo 💪

**Comment by stonko1994** [comment]
@younata I merged `main` into this branch to also use the transitive dependency changes 🙌 

I would still wait for https://github.com/Quick/Nimble/pull/1107 to be finished before I update the build on this PR.

**Review by younata — APPROVED** [review]

**Comment by younata** [comment]
@stonko1994 I (finally...) got around to getting #1107 working and merged. Looking forward to THAT being taken care of from now on. Please pull in current main, and I'll be happy to merge this as soon as I can.

File: Package.resolved
Comment (stonko1994) [inline Package.resolved:1]: This change to SPM version 2 should be fine, but up for discussion if we want to have it. Version 2 is available since Swift 5.6 which Xcode 13. I did a test opening the project with Xcode 14, and it worked without problems.

**Comment by stonko1994** [comment]
@younata I merged `main` into this branch 💪 Please execute the CI checks to see if everything works as expected now. Other than that it's ready to be merged from my side.

File: .github/workflows/ci-xcode.yml
Comment (younata) [inline .github/workflows/ci-xcode.yml:35]: Github actions doesn't yet support 15.2 - https://github.com/actions/runner-images/pull/9169

We'll see this test failing until that's merged.

**Review by younata — APPROVED** [review]
We'll wait a bit until Github Actions supports Xcode 15.2 before I merge, but this looks pretty good!

File: .github/workflows/ci-xcode.yml
Comment (stonko1994) [reply .github/workflows/ci-xcode.yml:35]: But it is mentioned [here](https://github.com/actions/runner-images/blob/releases/macos-13/20240114/images/macos/macos-13-Readme.md) 🤔

Comment (younata) [reply .github/workflows/ci-xcode.yml:35]: Oh. Well, excellent. Guess the thread I found was already out of date.

Comment (younata) [reply .github/workflows/ci-xcode.yml:35]: Well, that's annoying. Is the Vision Pro simulator not created by default? https://github.com/Quick/Nimble/actions/runs/7554289300/job/20585517470?pr=1098

Comment (stonko1994) [reply .github/workflows/ci-xcode.yml]: It looks like it 😞 

[6eb5881](https://github.com/Quick/Nimble/pull/1098/commits/6eb588189a4941efe9eaab83e60ba78aa5ee1c9e) I reverted it to unblock this PR. Will keep an eye on the GH runner images and add it once it changes.

Comment (stonko1994) [reply .github/workflows/ci-xcode.yml]: Just found [this issue](https://github.com/actions/runner-images/issues/9139) where they point out that the version of Xcode 15.2 is the beta 2 and not the final version. I guess it should be fine once they use the stable version of 15.2.

Comment (younata) [reply .github/workflows/ci-xcode.yml]: I'll watch that thread and update our CI once a new build is out.

**Pull request merged**
